pkg/ctrlmgr/controllers/cluster/cluster_controller.go

```diff
@@ -138,18 +164,50 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 	// start to scout loop for memberCluster warden, non-block
 	err = multicluster.Interface().ScoutFor(context.Background(), currentCluster.Name)
 	if err != nil {
-		log.Error("start scout for memberCluster %v failed", currentCluster.Name)
+		log.Error("start scout for cluster %v failed", currentCluster.Name)
+		return ctrl.Result{}, err
 	}
 
 	// update cluster status to processing
-	err = r.updateClusterStatus(ctx, currentCluster)
+	updateFn := func(cluster *clusterv1.Cluster) {
+		state := clusterv1.ClusterProcessing
+		reason := fmt.Sprintf("cluster %s initializing", cluster.Name)
+		cluster.Status.State = &state
+		cluster.Status.Reason = reason
+	}
+
+	err = utils.UpdateStatus(ctx, r.Client, &currentCluster, updateFn)
 	if err != nil {
-		log.Error("update cluster %v status failed", currentCluster.Name)
+		log.Warn("update cluster %v status failed: %v", currentCluster.Name, err)
 	}
 
 	return ctrl.Result{}, nil
 }
 
+// It enqueues a cluster for later reconciliation. This occurs in a goroutine
+// so the caller doesn't block; since the reconciler is never garbage-collected, this is safe.
+func (r *ClusterReconciler) enqueue(cluster clusterv1.Cluster) {
+	const retrySeconds = 7
+
+	config, _ := kubeconfig.LoadKubeConfigFromBytes(cluster.Spec.KubeConfig)
+
+	// try to reconnect with cluster api server, requeue if every is ok
+	go func() {
+		log.Info("cluster %v init failed, keep retry background", cluster.Name)
+		for {
+			select {
+			case <-time.Tick(retrySeconds * time.Second):
+				_, err := client.New(config, client.Options{Scheme: r.Scheme})
+				if err == nil {
+					log.Info("enqueuing cluster %v for reconciliation", cluster.Name)
+					r.Affected <- event.GenericEvent{Object: &cluster}
+					return
+				}
+			}
+		}
```

Comment:
for range time.Tick(7 * time.Second ) {
  do_some_thing()
}

Reply:
this is another way do retry